components/logins/src/db.rs

```diff
@@ -307,6 +308,57 @@ impl LoginDb {
         rows.collect::<Result<_>>()
     }
 
+    pub fn get_by_base_domain(&self, base_domain: &str) -> Result<Vec<Login>> {
+        // We just do a linear scan for now. Later we might choose to have an
+        // indexed reverse-host column or similar.
+        // Even less optimally, have sqlite return everything to us and filter
+        // in rust - we could probalby use a regex if we cared.
+        let mut stmt = self.db.prepare_cached(&GET_ALL_SQL)?;
+        let rows = stmt
+            .query_and_then(NO_PARAMS, Login::from_row)?
+            .filter(|r| {
+                match r {
+                    Ok(login) => {
+                        match Url::parse(&login.hostname) {
+                            Ok(u) => {
+                                match u.host() {
+                                    Some(Host::Domain(s)) => {
+                                        if s == base_domain {
+                                            true
+                                        } else {
+                                            // a fairly long-winded way of saying
+                                            // `login.hostname.ends_with('.' + base_domain);`
+                                            let mut rev_input = base_domain.chars().rev();
+                                            let mut rev_host = s.chars().rev();
+                                            loop {
+                                                match (rev_input.next(), rev_host.next()) {
+                                                    (Some(ref a), Some(ref b)) if a == b => {
+                                                        continue
+                                                    }
+                                                    (None, None) => {
+                                                        unreachable!("already checked equality!")
+                                                    }
+                                                    (None, Some(ref h)) => return *h == '.',
+                                                    _ => return false,
+                                                }
+                                            }
+                                        }
+                                    }
+                                    // ip addresses (presumably) should just match exactly?
+                                    Some(Host::Ipv4(a)) => a.to_string() == base_domain,
+                                    Some(Host::Ipv6(a)) => a.to_string() == base_domain,
+                                    _ => false,
```

Comment:
I would be tempted to return `false` for all such records, and to explicitly error out if the caller gave us something that is not a well-formed domain name.

Reply:
Wouldn't that mean that clients wouldn't suggest a login for an IP address even when a saved password explicitly exists for that IP address?

Reply:
Hrm, maybe; I guess we'd need to see the calling code to know for sure. But in that case maybe we should fall back to a literal string match for any record that doesn't have a domain, rather than special-casing for IP addresses.